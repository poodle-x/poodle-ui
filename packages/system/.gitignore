# List build files
/lib/
/mjs/


# Misc
/coverage/
node_modules
.cache/
/scriptsBuild/
.DS_Store
yarn-error.log
