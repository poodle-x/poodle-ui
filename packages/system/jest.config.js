// For a detailed explanation regarding each configuration property, visit:
// https://jestjs.io/docs/en/configuration.html

module.exports = {
	coverageDirectory: "coverage",
	testEnvironment: "jsdom",
	transform: {
		"^.+\\.(js|jsx|tsx|ts)$": "babel-jest",
	},
	testMatch: ["<rootDir>/src/**/?(*.)+(spec|test).(ts|js)?(x)"],
	moduleFileExtensions: ["js", "ts", "tsx"],
	setupFilesAfterEnv: ["<rootDir>/jest.setup.js"],
};
