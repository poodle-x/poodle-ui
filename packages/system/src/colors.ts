import * as CSS from "csstype";
import {
	ThemeValue,
	SystemThemeConfig,
	getThemeValue,
	ResponsiveProp,
	getMode,
	SystemConfigList,
} from "./";

export type ColorValue =
	| string
	| {
			[key: string]: string;
	  };

export interface ColorsThemeConfig {
	sets?: {
		[key: string]: ColorValue;
	};
	modes?: {
		[key: string]: {
			[key: string]: ColorValue;
		};
	};
}

export function transformColor(
	props: { [key: string]: any },
	value: ThemeValue<string>
) {
	return getColor(value)(props);
}

function getColorValue(
	sets: {
		[key: string]: ColorValue;
	},
	value: string
) {
	let base = "base";

	let key = value;

	if (value.indexOf(".") !== -1) {
		const parts = value.split(".");
		base = parts[1] || "base";
		key = parts[0];
	}

	const colorValue = sets[key] || {};

	if (typeof colorValue === "string") {
		return colorValue;
	}

	return colorValue[base] || "";
}

export function getColor(value: ThemeValue<string>) {
	return (props: { [key: string]: any; theme?: SystemThemeConfig }) => {
		const { theme } = props;

		const calcValue = getThemeValue<string>(props, value, "");

		const mode = getMode(props);

		const color = getColorValue(theme?.colors?.sets || {}, calcValue);

		const colorMode = getColorValue(
			theme?.colors?.modes?.[mode] || {},
			calcValue
		);

		return colorMode || color || calcValue;
	};
}

export interface ColorsProps {
	/**
	 * The CSS `background` property
	 */
	bg?: ResponsiveProp<CSS.Property.Background>;
	/**
	 * The CSS `background` property
	 */
	background?: ResponsiveProp<CSS.Property.Background>;
	/**
	 * The CSS `background-color` property
	 */
	bgc?: ResponsiveProp<CSS.Property.BackgroundColor>;
	/**
	 * The CSS `background-color` property
	 */
	backGroundColor?: ResponsiveProp<CSS.Property.BackgroundColor>;
	/**
	 * The CSS `color` property
	 */
	color?: ResponsiveProp<string | CSS.Property.Color>;
	/**
	 * The CSS `fill` property
	 */
	fill?: ResponsiveProp<CSS.Property.Fill>;
	/**
	 * The CSS `opacity` property
	 */
	opacity?: ResponsiveProp<CSS.Property.Fill>;
}

export const colorsSystem: SystemConfigList = {
	"bg,background": {
		properties: ["background"],
		transform: transformColor,
	},
	"bgc,backgroundColor": {
		properties: ["backgroundColor"],
		transform: transformColor,
	},
	color: {
		properties: ["color"],
		transform: transformColor,
	},
	fill: {
		properties: ["fill"],
		transform: transformColor,
	},
	opacity: true,
};
