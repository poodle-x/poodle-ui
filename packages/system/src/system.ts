import { getBreakpoint, getBreakpointsScale } from "./breakpoints";
import { getThemeValue } from "./";

/**
 * Custom function to override get media query function when the value passing
 * is a responsive type.
 */
export type MediaFn = (option: MediaFnOptions) => string;
export type GetBreakpointScale = (props: { [key: string]: any }) => string[];

export type SystemConfigTransform = (
	props: { [key: string]: any },
	value: any,
	key: string,
	property: string
) => string | number | { [key: string]: string | number } | undefined;

export type SystemConfigTransformProperties = (data: {
	props: { [key: string]: any };
	key: string;
}) => string[];

export interface SystemConfig {
	properties?: string[] | SystemConfigTransformProperties;
	transform?: SystemConfigTransform;
	customMediaFn?: MediaFn;
	customGetBreakpointsScale?: GetBreakpointScale;
	isMultipleValue?: boolean;
}

export interface MediaFnOptions {
	scale: number | string;
	value: number | string;
	props: { [key: string]: any };
	config: SystemConfig;
}

export interface CreateSystemOptions {
	customMediaFn?: MediaFn;
	customGetBreakpointsScale?: GetBreakpointScale;
}

export interface SystemConfigList {
	[key: string]: SystemConfig | true;
}

function getValue(data: {
	key: string;
	props: { [key: string]: any };
	config: SystemConfig;
	value: any;
}) {
	const { props, config, value, key } = data;

	let style: { [key: string]: any } = {};

	if (config.properties) {
		let rawProperties: string[] | undefined;

		if (typeof config.properties === "function") {
			rawProperties = config.properties({
				props,
				key,
			});
		} else {
			rawProperties = config.properties;
		}

		rawProperties.forEach((property) => {
			style[property] = config.transform
				? config.transform(props, value, key, property)
				: getThemeValue<string | number>(props, value, "");
		});
	} else {
		const v = config.transform
			? config.transform(props, value, key, key)
			: getThemeValue<{ [key: string]: string | number } | string | number>(
					props,
					value,
					{}
			  );

		if (typeof v === "string" || typeof v === "number") {
			style[key] = v;
		}

		if (typeof v === "object") {
			style = {
				...style,
				...v,
			};
		}
	}

	return style;
}

function mediaFn(options: MediaFnOptions): string {
	let media = "";
	const { scale, props } = options;
	const breakpoint = getBreakpoint(scale)(props);

	if (breakpoint) {
		media = `@media screen and (min-width: ${
			typeof breakpoint === "number" ? `${breakpoint}px` : breakpoint
		})`;
	}
	return media;
}

function mediaScaleFn(options: {
	key: string;
	config: SystemConfig;
	mediaFn: MediaFn;
	scale: string | number;
	styleValue: any;
	props: { [key: string]: any };
}) {
	const { config, scale, mediaFn, styleValue, props, key } = options;
	const media = mediaFn({
		config,
		scale,
		props,
		value: styleValue,
	});
	if (media) {
		return {
			[media]: getValue({
				key,
				config,
				value: styleValue,
				props,
			}),
		};
	}
}

function getKeyStyles(data: {
	key: string;
	config: SystemConfig;
	props: { [key: string]: any };
	sendProps?: { [key: string]: any };
	customMediaFn?: MediaFn;
	customGetBreakpointsScale?: GetBreakpointScale;
}): Array<{ [p: string]: any }> {
	const {
		config,
		props,
		key,
		customMediaFn,
		customGetBreakpointsScale,
		sendProps,
	} = data;
	const styles: Array<{ [key: string]: any }> = [];
	const currentValue = props[key];
	const currentMediaFn = config.customMediaFn || customMediaFn || mediaFn;
	const currentGetBreakpointsScale =
		config.customGetBreakpointsScale ||
		customGetBreakpointsScale ||
		getBreakpointsScale;

	const forwardProps = sendProps || props;

	if (Array.isArray(currentValue)) {
		currentValue.forEach((value, index) => {
			const keyValue = getValue({ config, value, props: forwardProps, key });
			if (index === 0) {
				styles.push(keyValue);
			} else {
				const media = currentMediaFn({
					config,
					scale: index - 1,
					props: forwardProps,
					value,
				});
				if (media) {
					styles.push({
						[media]: keyValue,
					});
				}
			}
		});
		return styles;
	}

	if (typeof currentValue === "object") {
		const listResponsive = Object.keys(currentValue);

		const spare = listResponsive.findIndex((v) => v === "_");

		if (spare >= 0) {
			styles.push(
				getValue({
					key,
					config,
					value: currentValue[listResponsive[spare]],
					props: forwardProps,
				})
			);
			listResponsive.splice(spare, 1);
		}

		const responsiveScale = currentGetBreakpointsScale(props);

		responsiveScale.forEach((scale) => {
			const spare = listResponsive.findIndex((v) => v === scale);
			if (spare >= 0) {
				const styleScale = mediaScaleFn({
					key,
					props: forwardProps,
					config,
					scale: listResponsive[spare],
					mediaFn: currentMediaFn,
					styleValue: currentValue[listResponsive[spare]],
				});
				styleScale && styles.push(styleScale);
				listResponsive.splice(spare, 1);
			}
		});

		listResponsive.forEach((res) => {
			const styleScale = mediaScaleFn({
				key,
				props: forwardProps,
				config,
				scale: res,
				mediaFn: currentMediaFn,
				styleValue: currentValue[res],
			});
			styleScale && styles.push(styleScale);
		});

		return styles;
	}

	if (
		typeof currentValue === "string" ||
		typeof currentValue === "number" ||
		typeof currentValue === "function"
	) {
		styles.push(
			getValue({ config, value: currentValue, props: forwardProps, key })
		);
		return styles;
	}

	return styles;
}

export interface SystemOperationOptions {
	returnOnlySystemProps?: boolean;
	sendProps?: { [key: string]: any };
}

export type SystemOperation = (
	props: { [p: string]: any },
	runOptions?: SystemOperationOptions
) => Array<{ [p: string]: any }>;

export function createSystem(
	systems: SystemConfigList,
	options?: CreateSystemOptions
): SystemOperation {
	const operation: SystemOperation = (props, runOptions = {}) => {
		const { returnOnlySystemProps, sendProps } = runOptions;
		const { customMediaFn, customGetBreakpointsScale } = options || {};
		let styles: Array<{ [key: string]: any }> = [];

		Object.keys(props)
			.filter((p) => p !== "theme")
			.forEach((p) => {
				let foundedInSystem = false;
				for (const keys in systems) {
					const keysPart = keys.split(",");

					const rawConfig = systems[keys];

					const key = keysPart.find((k) => k === p);

					foundedInSystem = Boolean(key);

					let config: SystemConfig;

					if (rawConfig === true) {
						config = {
							properties: [keys],
						};
					} else {
						config = rawConfig;
					}

					if (key) {
						if (config.isMultipleValue) {
							let rawProperties: string[] | undefined;

							if (typeof config.properties === "function") {
								rawProperties = config.properties({
									props,
									key,
								});
							} else {
								rawProperties = config.properties;
							}

							const properties = rawProperties || [key];

							properties.forEach((property) => {
								let value: { [key: string]: any } = props[key];

								if (config.transform) {
									const resovle = config.transform(
										props,
										props[key],
										key,
										property
									);

									if (
										resovle &&
										typeof resovle !== "number" &&
										typeof resovle !== "string"
									) {
										value = resovle;
									}
								}

								styles.push({
									[property]: operation(
										{ ...value, theme: props.theme },
										runOptions
									),
								});
							});
						} else {
							styles = [
								...styles,
								...getKeyStyles({
									key,
									config,
									props,
									sendProps,
									customMediaFn,
									customGetBreakpointsScale,
								}),
							];
						}
						break;
					}
				}

				if (
					!foundedInSystem &&
					!returnOnlySystemProps &&
					props[p] &&
					typeof props[p] === "object"
				) {
					return (styles = [
						...styles,
						{ [p]: operation({ ...props[p], theme: props.theme }, runOptions) },
					]);
				}

				if (!foundedInSystem && !returnOnlySystemProps) {
					styles = [...styles, { [p]: props[p] }];
				}
			});

		return styles;
	};

	return operation;
}

export function compose(...args: Array<SystemConfigList>): SystemConfigList {
	const finalList: SystemConfigList = {};
	args.forEach((list) => {
		Object.keys(list).forEach((key) => {
			const parts = key.split(",");
			parts.forEach((part) => {
				finalList[part] = list[key];
			});
		});
	});
	return finalList;
}
