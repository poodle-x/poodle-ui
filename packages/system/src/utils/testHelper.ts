import { SystemThemeConfig, createSystem, SystemConfigList } from "../";

export interface SystemTestCase {
	snapshot?: boolean;
	name: string;
	data: {
		theme?: SystemThemeConfig;
		steps: Array<Array<any>>;
	};
}

export function systemPropTestCases(data: {
	systemConfigs: SystemConfigList;
	testCases: SystemTestCase[];
	theme?: SystemThemeConfig;
}) {
	const { systemConfigs, testCases, theme } = data;
	testCases.forEach((tes) => {
		it(tes.name, () => {
			tes.data.steps.forEach((step) => {
				const result = createSystem(systemConfigs)({
					theme: {
						...theme,
						...tes.data.theme,
						...step[2],
					},
					...step[0],
				});
				if (tes.snapshot) {
					expect(result).toMatchSnapshot();
				} else {
					expect(result).toEqual(step[1]);
				}
			});
		});
	});
}
