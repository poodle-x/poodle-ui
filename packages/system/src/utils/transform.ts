import {
	SystemConfigTransform,
	SystemConfigTransformProperties,
} from "../system";

export function transformPropertyMultipleValue(
	part: string
): SystemConfigTransformProperties {
	return (data) => {
		const properties: string[] = [];

		const { props, key } = data;

		for (const k in props[key]) {
			if (props[key].hasOwnProperty(k)) {
				properties.push(`&:${part}(${k})`);
			}
		}

		return properties;
	};
}

export const transformMultipleValue: SystemConfigTransform = (
	props,
	value,
	key,
	property
) => {
	// @todo should change regex to sub string to increase performance
	const regex = /\((.+)\)/gm;

	const result = regex.exec(property);

	const keyValue: string = result ? result[1] : "";

	return value[keyValue];
};
