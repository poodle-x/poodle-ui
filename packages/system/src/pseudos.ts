import { SystemConfig } from "./system";
import {
	transformMultipleValue,
	transformPropertyMultipleValue,
} from "./utils/transform";

export const pseudosSystem: { [key: string]: SystemConfig } = {
	["_active"]: {
		properties: ["&:active"],
		isMultipleValue: true,
	},
	["_checked"]: {
		properties: ["&:checked"],
		isMultipleValue: true,
	},
	["_default"]: {
		properties: ["&:default"],
		isMultipleValue: true,
	},
	["_defined"]: {
		properties: ["&:defined"],
		isMultipleValue: true,
	},
	["_disabled"]: {
		properties: ["&:disabled"],
		isMultipleValue: true,
	},
	["_enabled"]: {
		properties: ["&:enabled"],
		isMultipleValue: true,
	},
	["_firstChild"]: {
		properties: ["&:first-child"],
		isMultipleValue: true,
	},
	["_firstOfType"]: {
		properties: ["&:first-of-type"],
		isMultipleValue: true,
	},
	["_focus"]: {
		properties: ["&:focus"],
		isMultipleValue: true,
	},
	["_focusWithin"]: {
		properties: ["&:focus-within"],
		isMultipleValue: true,
	},
	["_hover"]: {
		properties: ["&:hover"],
		isMultipleValue: true,
	},
	["_indeterminate"]: {
		properties: ["&:indeterminate"],
		isMultipleValue: true,
	},
	["_invalid"]: {
		properties: ["&:invalid"],
		isMultipleValue: true,
	},
	["_lang"]: {
		properties: transformPropertyMultipleValue("lang"),
		isMultipleValue: true,
		transform: transformMultipleValue,
	},
	["_lastChild"]: {
		properties: ["&:last-child"],
		isMultipleValue: true,
	},
	["_lastOfType"]: {
		properties: ["&:last-of-type"],
		isMultipleValue: true,
	},
	["_link"]: {
		properties: ["&:link"],
		isMultipleValue: true,
	},
	["_not"]: {
		properties: transformPropertyMultipleValue("not"),
		isMultipleValue: true,
		transform: transformMultipleValue,
	},
	["_nthChild"]: {
		properties: transformPropertyMultipleValue("nth-child"),
		isMultipleValue: true,
		transform: transformMultipleValue,
	},
	["_nthLastChild"]: {
		properties: transformPropertyMultipleValue("nth-last-child"),
		isMultipleValue: true,
		transform: transformMultipleValue,
	},
	["_nthLastOfType"]: {
		properties: transformPropertyMultipleValue("nth-last-of-type"),
		isMultipleValue: true,
		transform: transformMultipleValue,
	},
	["_nthOfType"]: {
		properties: transformPropertyMultipleValue("nth-of-type"),
		isMultipleValue: true,
		transform: transformMultipleValue,
	},
	["_onlyChild"]: {
		properties: ["&:only-child"],
		isMultipleValue: true,
	},
	["_onlyOfType"]: {
		properties: ["&:only-of-type"],
		isMultipleValue: true,
	},
	["_optional"]: {
		properties: ["&:optional"],
		isMultipleValue: true,
	},
	["_placeholderShown"]: {
		properties: ["&:placeholder-shown"],
		isMultipleValue: true,
	},
	["_readonly"]: {
		properties: ["&:read-only"],
		isMultipleValue: true,
	},
	["_readwrite"]: {
		properties: ["&:read-write"],
		isMultipleValue: true,
	},
	["_required"]: {
		properties: ["&:required"],
		isMultipleValue: true,
	},
	["_target"]: {
		properties: ["&:target"],
		isMultipleValue: true,
	},
	["_valid "]: {
		properties: ["&:valid"],
		isMultipleValue: true,
	},
	["_visited "]: {
		properties: ["&:visited"],
		isMultipleValue: true,
	},
	["_after"]: {
		properties: ["&::after"],
		isMultipleValue: true,
	},
	["_before"]: {
		properties: ["&::before"],
		isMultipleValue: true,
	},
	["_backdrop"]: {
		properties: ["&::backdrop"],
		isMultipleValue: true,
	},
	["_firstLetter"]: {
		properties: ["&::first-letter"],
		isMultipleValue: true,
	},
	["_firstLine"]: {
		properties: ["&::first-line"],
		isMultipleValue: true,
	},
	["_placeholder"]: {
		properties: ["&::placeholder"],
		isMultipleValue: true,
	},
	["_selection"]: {
		properties: ["&::selection"],
		isMultipleValue: true,
	},
};

export interface PseudosProps<T> {
	/**
	 * CSS pseudo-class :active
	 */
	_active?: T;
	/**
	 * CSS pseudo-class :checked
	 */
	_checked?: T;
	/**
	 * CSS pseudo-class :default
	 */
	_default?: T;
	/**
	 * CSS pseudo-class :defined
	 */
	_defined?: T;
	/**
	 * CSS pseudo-class :disabled
	 */
	_disabled?: T;
	/**
	 * CSS pseudo-class :enabled
	 */
	_enabled?: T;
	/**
	 * CSS pseudo-class :first-child
	 */
	_firstChild?: T;
	/**
	 * CSS pseudo-class :first-of-type
	 */
	_firstOfType?: T;
	/**
	 * CSS pseudo-class :focus
	 */
	_focus?: T;
	/**
	 * CSS pseudo-class :focus-within
	 */
	_focusWithin?: T;
	/**
	 * CSS pseudo-class :hover
	 */
	_hover?: T;
	/**
	 * CSS pseudo-class :indeterminate
	 */
	_indeterminate?: T;
	/**
	 * CSS pseudo-class :invalid
	 */
	_invalid?: T;
	/**
	 * CSS pseudo-class :lang()
	 */
	_lang?: T;
	/**
	 * CSS pseudo-class :last-child
	 */
	_lastChild?: T;
	/**
	 * CSS pseudo-class :last-of-type
	 */
	_lastOfType?: T;
	/**
	 * CSS pseudo-class :link
	 */
	_link?: T;
	/**
	 * CSS pseudo-class :not()
	 */
	_not?: T;
	/**
	 * CSS pseudo-class :nth-child()
	 */
	_nthChild?: T;
	/**
	 * CSS pseudo-class :nth-last-child()
	 */
	_nthLastChild?: T;
	/**
	 * CSS pseudo-class :nth-last-of-type()
	 */
	_nthLastOfType?: T;
	/**
	 * CSS pseudo-class :nth-of-type()
	 */
	_nthOfType?: T;
	/**
	 * CSS pseudo-class :only-child
	 */
	_onlyChild?: T;
	/**
	 * CSS pseudo-class :only-of-type
	 */
	_onlyOfType?: T;
	/**
	 * CSS pseudo-class :optional
	 */
	_optional?: T;
	/**
	 * CSS pseudo-class :placeholder-shown
	 */
	_placeholderShown?: T;
	/**
	 * CSS pseudo-class :readonly
	 */
	_readonly?: T;
	/**
	 * CSS pseudo-class :read-write
	 */
	_readwrite?: T;
	/**
	 * CSS pseudo-class :required
	 */
	_required?: T;
	/**
	 * CSS pseudo-class :target
	 */
	_target?: T;
	/**
	 * CSS pseudo-class :valid
	 */
	_valid?: T;
	/**
	 * CSS pseudo-class :visited
	 */
	_visited?: T;
	/**
	 * CSS pseudo-element :after
	 */
	_after?: T;
	/**
	 * CSS pseudo-element ::before
	 */
	_before?: T;
	/**
	 * CSS pseudo-element ::backdrop
	 */
	_backdrop?: T;
	/**
	 * CSS pseudo-element ::first-letter
	 */
	_firstLetter?: T;
	/**
	 * CSS pseudo-element ::placeholder
	 */
	_placeholder?: T;
	/**
	 * CSS pseudo-element ::selection
	 */
	_selection?: T;
}
