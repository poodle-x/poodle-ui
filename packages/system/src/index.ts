import { math } from "polished";
import { BreakpointsThemeConfig } from "./breakpoints";
import { ColorsThemeConfig } from "./colors";
import { SpacingThemeConfig } from "./spacing";
import { FontsThemeConfig } from "./fonts";
import { BordersThemeConfig } from "./borders";
import { LayoutsThemeConfig } from "./layout";
import { ShadowThemeConfig } from "./shadow";

export * from "./system";
export * from "./breakpoints";
export * from "./spacing";
export * from "./colors";
export * from "./layout";
export * from "./fonts";
export * from "./flex";
export * from "./grid";
export * from "./borders";
export * from "./pseudos";
export * from "./shadow";
export * from "./list";

export interface SystemThemeConfig {
	mode?: string;
	breakpoints?: BreakpointsThemeConfig;
	spacing?: SpacingThemeConfig;
	colors?: ColorsThemeConfig;
	fonts?: FontsThemeConfig;
	borders?: BordersThemeConfig;
	layouts?: LayoutsThemeConfig;
	shadows?: ShadowThemeConfig;
}

export type ThemeValue<T> = T | ((props: { [key: string]: any }) => T);

export type ResponsiveProp<T = any> =
	| ThemeValue<T>
	| {
			[key: string]: ThemeValue<T> | Array<ThemeValue<T>>;
	  }
	| ThemeValue<T>[];

export function getThemeValue<T>(
	props: { [key: string]: any },
	value: ThemeValue<T> | undefined,
	defaultValue: T
) {
	if (value === undefined) {
		return defaultValue;
	}

	if (typeof value === "function") {
		return (value as (props: { [key: string]: any }) => T)(props);
	}

	return value;
}

/**
 * `getScaling` return spacing value from received value and scaling value.
 *
 * If the value is a string like "scale-x" with 'x' is a number then the function
 * will return the result equal by scale multiply with 'x'.
 * If the value type is a string and is found in `sets[value]`, the function will use that and
 * return the result via `getThemeValue` function.
 * Other cases return the result via `getThemeValue` function with value is the same received value.
 */
export function getScaling(data: {
	props: { [key: string]: any };
	scale: number | string;
	value: ThemeValue<number | string>;
	sets?: {
		[k: string]: ThemeValue<number | string>;
	};
}) {
	const { scale, sets, value, props } = data;

	if (typeof value === "number") {
		return value;
	}

	let valueFinal: ThemeValue<number | string> = getThemeValue<string | number>(
		props,
		value,
		""
	);

	if (typeof valueFinal === "string") {
		const parts = valueFinal.split("-");

		const scaleNumber = parseFloat(parts[1]);

		if (parts[0] === "scale" && !isNaN(scaleNumber)) {
			return typeof scale === "string"
				? math(`${scaleNumber} * ${scale}`)
				: scaleNumber * scale;
		}
		valueFinal = sets?.[valueFinal] || valueFinal;
	}

	return getThemeValue<string | number>(props, valueFinal, "");
}

export function getMode(props: {
	[key: string]: any;
	theme?: SystemThemeConfig;
}) {
	const { theme, mode } = props;
	if (typeof mode === "string") {
		return mode;
	}
	return theme?.mode || "";
}
