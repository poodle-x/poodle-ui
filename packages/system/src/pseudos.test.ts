import { pseudosSystem } from "./pseudos";
import { systemPropTestCases, SystemTestCase } from "./utils/testHelper";
import { compose } from "./system";
import { colorsSystem } from "./colors";

describe("pseudos system", function () {
	describe("pseudos props", function () {
		const testCases: SystemTestCase[] = [
			{
				name: "_active props",
				snapshot: true,
				data: {
					steps: [
						[
							{
								_active: {
									color: ["primary", "secondary"],
								},
							},
						],
					],
				},
			},
			{
				name: "_checked props",
				snapshot: true,
				data: {
					steps: [
						[
							{
								_checked: {
									color: ["primary", "secondary"],
								},
							},
						],
					],
				},
			},
			{
				name: "_default props",
				snapshot: true,
				data: {
					steps: [
						[
							{
								_default: {
									color: ["primary", "secondary"],
								},
							},
						],
					],
				},
			},
			{
				name: "_defined props",
				snapshot: true,
				data: {
					steps: [
						[
							{
								_defined: {
									color: ["primary", "secondary"],
								},
							},
						],
					],
				},
			},
			{
				name: "_disabled props",
				snapshot: true,
				data: {
					steps: [
						[
							{
								_disabled: {
									color: ["primary", "secondary"],
								},
							},
						],
					],
				},
			},
			{
				name: "_enabled props",
				snapshot: true,
				data: {
					steps: [
						[
							{
								_enabled: {
									color: ["primary", "secondary"],
								},
							},
						],
					],
				},
			},
			{
				name: "_firstChild props",
				snapshot: true,
				data: {
					steps: [
						[
							{
								_firstChild: {
									color: ["primary", "secondary"],
								},
							},
						],
					],
				},
			},
			{
				name: "_firstOfType props",
				snapshot: true,
				data: {
					steps: [
						[
							{
								_firstOfType: {
									color: ["primary", "secondary"],
								},
							},
						],
					],
				},
			},
			{
				name: "_focus props",
				snapshot: true,
				data: {
					steps: [
						[
							{
								_focus: {
									color: ["primary", "secondary"],
								},
							},
						],
					],
				},
			},
			{
				name: "_focusWithin props",
				snapshot: true,
				data: {
					steps: [
						[
							{
								_focusWithin: {
									color: ["primary", "secondary"],
								},
							},
						],
					],
				},
			},
			{
				name: "_hover props",
				snapshot: true,
				data: {
					steps: [
						[
							{
								_hover: {
									color: ["primary", "secondary"],
								},
							},
						],
					],
				},
			},
			{
				name: "_indeterminate props",
				snapshot: true,
				data: {
					steps: [
						[
							{
								_indeterminate: {
									color: ["primary", "secondary"],
								},
							},
						],
					],
				},
			},
			{
				name: "_invalid props",
				snapshot: true,
				data: {
					steps: [
						[
							{
								_invalid: {
									color: ["primary", "secondary"],
								},
							},
						],
					],
				},
			},
			{
				name: "_lang props",
				snapshot: true,
				data: {
					steps: [
						[
							{
								en: {
									color: ["red", "blue"],
									fontFamily: "en",
								},
								zh: {
									color: ["cyan", "purple"],
									fontFamily: "cn",
								},
							},
						],
					],
				},
			},
			{
				name: "_lastChild props",
				snapshot: true,
				data: {
					steps: [
						[
							{
								_lastChild: {
									color: ["primary", "secondary"],
								},
							},
						],
					],
				},
			},
			{
				name: "_lastOfType props",
				snapshot: true,
				data: {
					steps: [
						[
							{
								_lastOfType: {
									color: ["primary", "secondary"],
								},
							},
						],
					],
				},
			},
			{
				name: "_lastOfType props",
				snapshot: true,
				data: {
					steps: [
						[
							{
								_lastOfType: {
									color: ["primary", "secondary"],
								},
							},
						],
					],
				},
			},
		];

		systemPropTestCases({
			testCases,
			systemConfigs: compose(pseudosSystem, colorsSystem),
			theme: {
				breakpoints: {
					scale: ["sm", "md", "lg", "xl"],
					sets: {
						sm: 768,
						md: 1024,
						lg: 1280,
						xl: 1600,
					},
				},
				colors: {
					sets: {
						primary: "red",
						secondary: "yellow",
					},
					modes: {
						dark: {
							primary: "cyan",
						},
					},
				},
			},
		});
	});
});
