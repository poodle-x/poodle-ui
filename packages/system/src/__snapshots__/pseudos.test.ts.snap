// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pseudos system pseudos props _active props 1`] = `
Array [
  Object {
    "&:active": Array [
      Object {
        "color": "red",
      },
      Object {
        "@media screen and (min-width: 768px)": Object {
          "color": "yellow",
        },
      },
    ],
  },
]
`;

exports[`pseudos system pseudos props _checked props 1`] = `
Array [
  Object {
    "&:checked": Array [
      Object {
        "color": "red",
      },
      Object {
        "@media screen and (min-width: 768px)": Object {
          "color": "yellow",
        },
      },
    ],
  },
]
`;

exports[`pseudos system pseudos props _default props 1`] = `
Array [
  Object {
    "&:default": Array [
      Object {
        "color": "red",
      },
      Object {
        "@media screen and (min-width: 768px)": Object {
          "color": "yellow",
        },
      },
    ],
  },
]
`;

exports[`pseudos system pseudos props _defined props 1`] = `
Array [
  Object {
    "&:defined": Array [
      Object {
        "color": "red",
      },
      Object {
        "@media screen and (min-width: 768px)": Object {
          "color": "yellow",
        },
      },
    ],
  },
]
`;

exports[`pseudos system pseudos props _disabled props 1`] = `
Array [
  Object {
    "&:disabled": Array [
      Object {
        "color": "red",
      },
      Object {
        "@media screen and (min-width: 768px)": Object {
          "color": "yellow",
        },
      },
    ],
  },
]
`;

exports[`pseudos system pseudos props _enabled props 1`] = `
Array [
  Object {
    "&:enabled": Array [
      Object {
        "color": "red",
      },
      Object {
        "@media screen and (min-width: 768px)": Object {
          "color": "yellow",
        },
      },
    ],
  },
]
`;

exports[`pseudos system pseudos props _firstChild props 1`] = `
Array [
  Object {
    "&:first-child": Array [
      Object {
        "color": "red",
      },
      Object {
        "@media screen and (min-width: 768px)": Object {
          "color": "yellow",
        },
      },
    ],
  },
]
`;

exports[`pseudos system pseudos props _firstOfType props 1`] = `
Array [
  Object {
    "&:first-of-type": Array [
      Object {
        "color": "red",
      },
      Object {
        "@media screen and (min-width: 768px)": Object {
          "color": "yellow",
        },
      },
    ],
  },
]
`;

exports[`pseudos system pseudos props _focus props 1`] = `
Array [
  Object {
    "&:focus": Array [
      Object {
        "color": "red",
      },
      Object {
        "@media screen and (min-width: 768px)": Object {
          "color": "yellow",
        },
      },
    ],
  },
]
`;

exports[`pseudos system pseudos props _focusWithin props 1`] = `
Array [
  Object {
    "&:focus-within": Array [
      Object {
        "color": "red",
      },
      Object {
        "@media screen and (min-width: 768px)": Object {
          "color": "yellow",
        },
      },
    ],
  },
]
`;

exports[`pseudos system pseudos props _hover props 1`] = `
Array [
  Object {
    "&:hover": Array [
      Object {
        "color": "red",
      },
      Object {
        "@media screen and (min-width: 768px)": Object {
          "color": "yellow",
        },
      },
    ],
  },
]
`;

exports[`pseudos system pseudos props _indeterminate props 1`] = `
Array [
  Object {
    "&:indeterminate": Array [
      Object {
        "color": "red",
      },
      Object {
        "@media screen and (min-width: 768px)": Object {
          "color": "yellow",
        },
      },
    ],
  },
]
`;

exports[`pseudos system pseudos props _invalid props 1`] = `
Array [
  Object {
    "&:invalid": Array [
      Object {
        "color": "red",
      },
      Object {
        "@media screen and (min-width: 768px)": Object {
          "color": "yellow",
        },
      },
    ],
  },
]
`;

exports[`pseudos system pseudos props _lang props 1`] = `
Array [
  Object {
    "en": Array [
      Object {
        "color": "red",
      },
      Object {
        "@media screen and (min-width: 768px)": Object {
          "color": "blue",
        },
      },
      Object {
        "fontFamily": "en",
      },
    ],
  },
  Object {
    "zh": Array [
      Object {
        "color": "cyan",
      },
      Object {
        "@media screen and (min-width: 768px)": Object {
          "color": "purple",
        },
      },
      Object {
        "fontFamily": "cn",
      },
    ],
  },
]
`;

exports[`pseudos system pseudos props _lastChild props 1`] = `
Array [
  Object {
    "&:last-child": Array [
      Object {
        "color": "red",
      },
      Object {
        "@media screen and (min-width: 768px)": Object {
          "color": "yellow",
        },
      },
    ],
  },
]
`;

exports[`pseudos system pseudos props _lastOfType props 1`] = `
Array [
  Object {
    "&:last-of-type": Array [
      Object {
        "color": "red",
      },
      Object {
        "@media screen and (min-width: 768px)": Object {
          "color": "yellow",
        },
      },
    ],
  },
]
`;

exports[`pseudos system pseudos props _lastOfType props 2`] = `
Array [
  Object {
    "&:last-of-type": Array [
      Object {
        "color": "red",
      },
      Object {
        "@media screen and (min-width: 768px)": Object {
          "color": "yellow",
        },
      },
    ],
  },
]
`;
